Fix #475, #382: Fix lookup of actions when tag contains dash

bsieber-mozilla opened the pull request:
Resolves: #475 

Updates to lookup_action logic-- instead of constructing tags from the whiteboard, then trying to see which tag exists as an action. In this commit we attempt to use the action.whiteboard_tag as the element which must exist within the bug.whiteboard element--iterating through the actions, and acting on the first action that applies to the bug. The whiteboard_tag presented in the action must be contained within the bug’s whiteboard field with a prefix of [

As is: the functionality is working with this new process for action confirmation. 

- [x] Fix logs so that when ActionNotFound is raised and used later for processing, this makes sense in the messaging


This PR ultimately allows for [exact-match-or-partial-matchs-123]:
- `data-eng-platform` and `data-eng-quality` would be able to coexist and map to separate processes

bsieber-mozilla commented on tests/unit/test_runner.py:
```diff
@@ -58,6 +58,34 @@ def test_request_is_ignored_because_private(
     assert str(exc_info.value) == "private bugs are not valid for action 'devtest'"
 
 
+def test_request_(
```
```suggestion
def test_request_extended_whiteboard_match(
```

bsieber-mozilla commented on tests/unit/test_router.py:
```diff
@@ -125,7 +125,7 @@ def test_webhook_is_200_if_action_raises_IgnoreInvalidRequestError(
         assert (
             response.json()["error"]
-            == "no action matching bug whiteboard tags: unmatched"
+            == "no action matching bug whiteboard tags: devtest"
```
this log makes a little less sense with this transition

bsieber-mozilla replied:
```suggestion
            == "no whiteboard matching action tags: devtest"
```

leplatrem commented:
Bryan, could you explain the reason behind this change? 
It may be a good opportunity to start a `docs/adrs/` folder, and add a brief one for this. 

> Side nitpicking: there's no issue associated to this ;)

bsieber-mozilla commented:
> Bryan, could you explain the reason behind this change?
> It may be a good opportunity to start a docs/adrs/ folder, and add a brief one for this.

We have an action that currently is impossible to map whiteboard tags to with the current `.split` process that occurs on the `-` character: 
https://github.com/mozilla/jira-bugzilla-integration/blob/846d69a6d661065ac2fcfe802f8548ac34123d6b/config/config.prod.yaml#L221

Currently any whiteboard tag that comes in with: `[remote-settings]` gets assigned the action_tag `remote` as the lookup key in the action dictionary.

Reversing the lookup to find the action_tag inside the whiteboard_tag without parsing the whiteboard_tag as heavily; enables this action as well as future actions to have "-"s in them. 

In the later commits, I added `brackets_required` to the Action model; the default behavior of having the brackets required can now be disabled per action. This has been implicit and is now likely a process many users follow.

This PR now makes [this behavior] something we are not requiring, but are defaulting to at this time.

bsieber-mozilla commented:
>     Side nitpicking: there's no issue associated to this ;)

Ahh, upon second reading this made sense; I thought you were mentioning there's no issue with the code.. but there's no github issue associated with this change, is the conclusion I just came to... I made #475 from the DENG comment, forcing a user to adjust their whiteboard processing. 

I'll add the above comment in regards to [remote-settings] to the gh issue as well

grahamalama commented on tests/conftest.py:
```diff
@@ -225,11 +225,35 @@ def action_example() -> Action:
     return action_factory()
 
 
+@pytest.fixture
+def action_example_with_inner_match() -> Action:
+    return action_factory(whiteboard_tag="inner-match")
+
+
+@pytest.fixture
+def action_example_without_brackets_required() -> Action:
+    return action_factory(whiteboard_tag="bracket-less", brackets_required=False)
+
+
 @pytest.fixture
 def actions_example(action_example) -> Actions:
     return Actions.parse_obj([action_example])
 
 
+@pytest.fixture
+def actions_example_with_inner_match(
+    action_example_with_inner_match,
+) -> Actions:
+    return Actions.parse_obj([action_example_with_inner_match])
+
+
+@pytest.fixture
+def actions_example_without_brackets_required(
+    action_example_without_brackets_required,
+) -> Actions:
+    return Actions.parse_obj([action_example_without_brackets_required])
+
+
```
If these fixtures are only used in one or two tests, it may be better to define them at the module level, rather than clutter the root `conftest.py`.

Similarly, it looks like `action_example_with_inner_match` is only used to build `actions_example_with_inner_match`. Maybe we can just do

```python
@pytest.fixture
def actions_example_with_inner_match() -> Actions:
    action = action_factory(whiteboard_tag="inner-match")
    return Actions.parse_obj([action])
```

repeating that pattern for `action_example_without_brackets_required` / `actions_example_without_brackets_required`.

grahamalama replied:
I'm also noticing that these factories would be more useful if we had fixtures that simply returned the factory function, e.g. 

```
@pytest.fixture
def action_factory() -> Action:
    return factories.action_factory
```

Then we could use them directly in tests.

bsieber-mozilla replied:
```
@pytest.fixture
def action_factory() -> Action:
    return factories.action_factory
```

Initially there was some overloading issues when trying this out; but I modified the import practices to allow for this naming convention (and usage in the tests).

As for the repeating of tests, the newest commit is aimed to address that concern as well.

grahamalama commented on tests/unit/test_runner.py:
Should we be testing the changes of this PR somewhere in `tests/unit/test_bugzilla.py`, maybe similar to [these](https://github.com/mozilla/jira-bugzilla-integration/blob/044e4a21a1d8259b2c99ba2be13aa7844a4a1b64/tests/unit/test_bugzilla.py#L56-L67) tests? Testing at the runner level perhaps feels too "wide" to make assertions about when it seems like we're asserting specific behavior of how `BugzillaBug.lookup_action` works.

bsieber-mozilla replied:
I added some more tests to `test_bugzilla`; I agree `test_runner` might be a bit "wide" but I do believe the new tests coexist with some of the prior testing strategy/process in the `test_runner`

bsieber-mozilla commented on tests/conftest.py:
```diff
@@ -206,54 +206,37 @@ def webhook_change_status_assignee():
-def action_example_without_brackets_required() -> Action:
-    return action_factory(whiteboard_tag="bracket-less", brackets_required=False)
+def action_example() -> Action:
+    return factories.action_factory()
```
_with the usage of `action_factory` from the `factories` module being imported to the namespace; the name `action_factory` was not possible in this file prior._

ref: https://github.com/mozilla/jira-bugzilla-integration/pull/472#discussion_r1167237261

leplatrem commented on README.md:
```diff
@@ -34,9 +34,15 @@ A bit more about the different fields...
+- `brackets_required` (optional)
+    - bool [true, false]
+    - default: true
+    - If `true` the whiteboard_tag must be be contained within an opening ([) and closing (]) bracket, this is the default behavior; i.e. this tag `XYX` matches this whiteboard_tag `[XYZ]` but not `XYZ`.
```
Something that I figured out reading the code as well: it would match: `[xyz]`, `[something-XYZ]`, `[XYZ-something]`

leplatrem commented on tests/unit/test_bugzilla.py:
```diff
@@ -60,11 +61,43 @@ def test_lookup_action(actions_example):
+    bug = bug_factory(id=1234, whiteboard="[example][foo-DevTest-bar]")
+    action = bug.lookup_action(actions_example)
+    assert action.whiteboard_tag == "devtest"
+    assert "test config" in action.description
```
Since these 3 tests are constructed the same way, you can use `@pytest.parameterize` passing the whiteboard.

leplatrem commented on jbi/models.py:
```diff
@@ -304,11 +306,15 @@ def extract_from_see_also(self):
+            for tag, action in actions.by_tag.items():
+                pre = r"\[.*" if action.brackets_required else ""
+                post = r".*\]" if action.brackets_required else ""
+                search_string = f"{pre}{tag.lower()}{post}"
```
`[.*{tag}.*]` may be too permissive:

```
>>> import re
>>> re.search(r"\[.*devtest.*\]", "[foo] devtest [bar]")
<re.Match object; span=(0, 19), match='[foo] devtest [bar]'>
>>>
```

you would have to change it to something like this:

```
>>> re.search(r"\[[^\]]*devtest[^\]]*\]", "[foo] devtest [bar]")
>>>
>>> re.search(r"\[[^\]]*devtest[^\]]*\]", "[foo-devtest-bar]")
<re.Match object; span=(0, 17), match='[foo-devtest-bar]'>
>>>
```
≈

leplatrem commented on README.md:
```diff
@@ -34,9 +34,15 @@ A bit more about the different fields...
+    - If `false` bugs that are not public will not be synchronized. Note that in order to synchronize
       private bugs the bugzilla user that JBI runs as must be in the security groups that are making
       the bug private.
+- `brackets_required` (optional)
```
If this option really necessary? I ask because no configuration entry was modified in this PR

leplatrem commented on jbi/models.py:
```diff
@@ -304,11 +306,15 @@ def extract_from_see_also(self):
+                pre = r"\[.*" if action.brackets_required else ""
+                post = r".*\]" if action.brackets_required else ""
+                search_string = f"{pre}{tag.lower()}{post}"
+                if re.search(search_string, self.whiteboard.lower()):
```
nitpick: instead of turning both inputs to lowercase, you can do: `if re.search(search_string, self.whiteboard, flags=re.IGNORE_CASE):`
